### PR TITLE
Build/copy macOS frameworks to built products instead of ephemeral directory

### PR DESCRIPTION
**Pull request by jmagman**
## Description

macOS version of https://github.com/flutter/flutter/pull/69699 and part of https://github.com/flutter/flutter/pull/70224.

- Instead of building App.framework and copying FlutterMacOS.framework to `Flutter/ephemeral`, instead target `BUILT_PRODUCTS_DIR` (for example `build/macos/Build/Products/Debug/`).  
- Remove `FRAMEWORK_SEARCH_PATHS` build setting from template and all example projects to prefer the built products directory, which we always get for free.  I didn't write a migrator since leaving this around in existing apps is harmless--the `inherited` part means it will find our copied framework first.

This:
1. Prevents issues like https://github.com/flutter/flutter/issues/37850 where `flutter build macos --debug` is run right before running in Release/archiving in Xcode and some pieces build before the xcconfig can be updated.
2. Makes caching indirectly aware of `$CONFIGURATION`s so we don't have to keep track of whether to rebuild or recopy frameworks when switching from Debug to Release.  Either the frameworks are in `build/macos/Build/Products/{Debug,Release}/` or they aren't.

## Related Issues
https://github.com/flutter/flutter/pull/72372
Blocked by https://github.com/flutter/flutter/pull/72020, which changes plugins from linking on `Flutter/ephemeral/FlutterMacOS.framework` to instead linking directly on the framework in the engine artifacts.

## Tests

Updated macos_test 
Already validated by [macos_content_validation_test](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/integration.shard/macos_content_validation_test.dart)

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_system/targets/macos.dart [hunk 1]
```diff
@@ -38,53 +38,36 @@ abstract class UnpackMacOS extends Target {
   ];
 
   @override
-  List<Source> get outputs => const <Source>[];
+  List<Source> get outputs => const <Source>[
+    Source.pattern('{OUTPUT_DIR}/FlutterMacOS.framework/FlutterMacOS'),
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/build_system/targets/macos.dart:42]: Either it's already been copied to `build/macos/Build/Products/Debug` or it hasn't, so we don't need `unpack_macos.d` anymore.   The subclasses include `Source.artifact(Artifact.flutterMacOSFramework, mode: x)` which will cause re-copies on flutter upgrades.

File: packages/flutter_tools/lib/src/build_system/targets/macos.dart [hunk 2]
```diff
@@ -38,53 +38,36 @@ abstract class UnpackMacOS extends Target {
...
+
+    final ProcessResult result = environment.processManager.runSync(<String>[
+      'rsync',
+      '-av',
+      '--delete',
+      '--filter',
+      '- .DS_Store/',
+      basePath,
+      environment.outputDir.path,
+    ]);
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/build_system/targets/macos.dart:64]: Please update the comment on this class, which currently says we're using `cp -R`.

**Review by stuartmorgan-g — APPROVED** [review]
LGTM!

**Comment by jmagman** [comment]
Google testing failed, but I don't see any test results for that PR in FRoB, failing or otherwise.  Since it passed before my comment-only change I'll call it good.

**Pull request merged**